Adding the collection contract and testing frameworks for this

kokeshiM0chi opened the pull request:
# Overview

This PR adds the following features:

-  flame work for testing contracts
-  contracts for testing and tests to check if `submessage` can be sent

This test is intended to be used in a CI environment.
However, it can be used in a local environment if you follow the instructions in operation.

## flame work for testing contracts

config and start the finschia-app docker used for testing.
- scripts/set_config_and_start_docker.sh

To organize the environment for this test, the osmosis script under scripts/ has been deleted.


## contracts for testing

- finschia execution environment using docker for the contract test
- wasm binary that built the contract
- Test about `wasm tx` of contract

Since stargate query does not yet be supported, we will not test it this time.
ref:
https://github.com/Finschia/finschia-wasm/issues/2

### examples/contracts/collection contract

A collection contract is Finschia's unique contract. Therefore, I am listing them as an example. Contracts need to be added if necessary.

tips:

<details><summary>Operation</summary><div>

Generate the wasm file by this command.

At first.

```
cargo build
```

Generate *.wasm files.

```
docker run --rm -v "$(pwd)":/code \
  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/examples/target \
  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
  cosmwasm/rust-optimizer:0.12.13 ./examples
```

App start:

```
cd scripts
./start_integration_test_contracts.sh
```


</div></details>

## related issue

The following issue was detected during contract testing.
https://github.com/Finschia/finschia-wasm/issues/4

kokeshiM0chi committed:
feat: add execution environment for finschia using docker

kokeshiM0chi committed:
feat: add the examle contracts

kokeshiM0chi committed:
feat: add integration contract tests

170210 commented on artifacts/checksums_intermediate.txt:
This file seems no need

170210 commented on artifacts/checksums.txt:
Since this file hasn't been used, I think it is no need.

loloicci replied:
I think this should not be deleted. Users can use checksum as like wasm's id.

170210 reviewed (commented):
Please use finschia/tutorial to build docker, it is no need to maintenance start.sh, run_finschia.sh, setup.sh in this repo.

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+chain_dir=${cur_path}"/finschia/template/.finschia"
+artifacts_path=${cur_path}"/../artifacts/"
+
+echo "hogehoge"
```
delete

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+sleep 30
+
+# send transaction to node
+cur_path=`pwd`
+chain_dir=${cur_path}"/finschia/template/.finschia"
+artifacts_path=${cur_path}"/../artifacts/"
```
I think it is better to do the test all in the docker.

kokeshiM0chi replied:
I want to be able to check it on local, so I'll leave it as is for now.

kokeshiM0chi replied:
>do the test all in the docker 

After considering it, I fixed it as pointed out in this commit.

[fix: change to send fnsad tx wasm execute ~ on docker instead of on…](https://github.com/Finschia/finschia-wasm/pull/3/commits/19b863ef9e0e09e959529308fbcba3b4568e2ef8)

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** store collection contract ***
+# store contract
+fnsad tx wasm store ${artifacts_path}/collection.wasm --node http://localhost:26658 --home=${chain_dir} --from link146asaycmtydq45kxc8evntqfgepagygelel00h --chain-id simd-testing --keyring-backend test --gas 10000000  -b block -o json -y > tmp.json
```
Need prepare const for 26658, link146asaycmtydq45kxc8evntqfgepagygelel00h

kokeshiM0chi replied:
I prepared const valuables in this commit. Please check it.

[fix: define the valuables with values used in this test](https://github.com/Finschia/finschia-wasm/pull/3/commits/cc1b47fe9a97c070cb48f2c3bb87050565d68b5d)

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# store contract
+fnsad tx wasm store ${artifacts_path}/collection.wasm --node http://localhost:26658 --home=${chain_dir} --from link146asaycmtydq45kxc8evntqfgepagygelel00h --chain-id simd-testing --keyring-backend test --gas 10000000  -b block -o json -y > tmp.json
+
+# confirm a result of `store`
+key=$(cat ./tmp.json | jq '.logs[] | .events' | jq '.[0]' | jq '.attributes[2]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events' | jq '.[0]' | jq '.attributes[2]' | jq '.value')
+expected_key="\"module\""
+expected_val="\"wasm\""
+
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
If you only need to check the field name, simply checking if 'failed' appears in the response is sufficient.

kokeshiM0chi replied:
Corrected so that it would terminate incorrectly in case of fail.

170210 reviewed (commented):
This PR appears to only test 4 WebAssembly files. Do you plan to test all other WebAssembly files in this PR?

da1suk8 commented on scripts/finschia/template/setup.sh:
```diff
@@ -0,0 +1,81 @@
+
+${FNSAD} validate-genesis --home=${CHAIN_DIR}
+
+# ${FNSAD} start --log_level *:debug --rpc.laddr=tcp://0.0.0.0:26657 --p2p.laddr=tcp://0.0.0.0:26656
```
Is this needed?

loloicci commented on scripts/finschia/env:
```diff
@@ -0,0 +1,4 @@
+# Choose from https://hub.docker.com/r/finschia/finschianode/tags
+REPOSITORY="finschia/finschianode"
+VERSION="2.0.0"
+CONTAINER_NAME="finschia-app"
```
add a blank line in the end

loloicci replied:
not fixed yet

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -0,0 +1,47 @@
+# If this error `realpath: command not found` occurs, run the following command.
+# brew install coreutils
+SCRIPT_DIR="$(realpath "$(dirname "$0")")"
+# shellcheck source=./env
+# shellcheck disable=SC1091
```
are these lines needed?

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -0,0 +1,47 @@
+source "$SCRIPT_DIR"/env
+
+TMP_DIR=$(mktemp -d "${TMPDIR:-/tmp}/fnsa.XXXXXXXXX")
+chmod 777 "$TMP_DIR"
```
isn't 755 enough?

loloicci commented on scripts/finschia/template/run_finschia.sh:
```diff
@@ -0,0 +1,7 @@
+
+cp -R "/template/.finschia" /root/.finschia
+mkdir -p /root/log
+fnsad start --rpc.laddr tcp://0.0.0.0:26658 --trace
```
add a blank line in the end

loloicci commented on scripts/finschia/template/setup.sh:
```diff
@@ -0,0 +1,81 @@
+mode="mainnet"
+
+SCRIPT_DIR="$(realpath "$(dirname "$0")")"
+# shellcheck source=./env
+# shellcheck disable=SC1091
```
are these needed?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash -e
+
+# sleep
```
is this line needed?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# sleep
+# This sleep is the sum of the time when docker is pulled, started,
+# and the height of the node increases to 1 or more.
+sleep 30
```
could you use other than simple sleep? For example, check the block height and retry after some sleep.

kokeshiM0chi replied:
Required as time to download docker.

kokeshiM0chi replied:
Fixed to wait until height becomes 1 or more

[fix: correct to wait for processing until height becomes 1 or more](https://github.com/Finschia/finschia-wasm/pull/3/commits/105bc0da23b62dd11851c972792742a88e4d8476)

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+# send transaction to node
+cur_path=`pwd`
+chain_dir=${cur_path}"/finschia/template/.finschia"
```
```suggestion
chain_dir="${cur_path}/finschia/template/.finschia"
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# send transaction to node
+cur_path=`pwd`
+chain_dir=${cur_path}"/finschia/template/.finschia"
+artifacts_path=${cur_path}"/../artifacts/"
```
```suggestion
artifacts_path="${cur_path}/../artifacts/"
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+artifacts_path=${cur_path}"/../artifacts/"
+
+echo "hogehoge"
+echo ${artifacts_path}
+echo ${chain_dir}
```
are these needed?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** store collection contract ***
+# store contract
+fnsad tx wasm store ${artifacts_path}/collection.wasm --node http://localhost:26658 --home=${chain_dir} --from link146asaycmtydq45kxc8evntqfgepagygelel00h --chain-id simd-testing --keyring-backend test --gas 10000000  -b block -o json -y > tmp.json
```
Is tmp.json needed? Isn't it enough to save it as a variable?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `store`
+key=$(cat ./tmp.json | jq '.logs[] | .events' | jq '.[0]' | jq '.attributes[2]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events' | jq '.[0]' | jq '.attributes[2]' | jq '.value')
+expected_key="\"module\""
+expected_val="\"wasm\""
```
```suggestion
expected_key='"module"'
expected_val='"wasm"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+expected_key="\"module\""
+expected_val="\"wasm\""
+
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
```suggestion
if [[ "${expected_key}" = "${key}" ]]; then echo "ok"; else echo "ng"; fi
if [[ "${expected_val}" = "${value}" ]]; then echo "ok"; else echo "ng"; fi
```

And, I doubt are these needed, too. At least, else should stop executing this script.

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** instantiate collection contract ***
+# instantiate contract
+fnsad tx wasm instantiate 1 '{"name":"collection_name","uri":"collection_uri","meta":"collection_meta", "owner":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"}' --label collection1 --admin link146asaycmtydq45kxc8evntqfgepagygelel00h  --from link146asaycmtydq45kxc8evntqfgepagygelel00h --node http://localhost:26658 --home=${chain_dir} --chain-id simd-testing --keyring-backend tes ...
```
Is tmp.json needed? Isn't it enough to save it as a variable?

And, delete a space before "-b block".

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `instantiate`
+key=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[1]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[1]' | jq '.value')
+expected_key="\"code_id\""
+expected_val="\"1\""
```
```suggestion
expected_key='"code_id"'
expected_val='"1"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[1]' | jq '.value')
+expected_key="\"code_id\""
+expected_val="\"1\""
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
same as L27-28

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** issue_nft collection contract ***
+# issue nft
+fnsad tx wasm execute link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8 '{"issue_nft":{"name":"nft1_name","meta":"nft1_meta","owner":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"}}' --from link146asaycmtydq45kxc8evntqfgepagygelel00h --node http://localhost:26658 --home=${chain_dir} --chain-id simd-testing --keyring-backend test --gas 10000000 -b block -o json -y  ...
```
Is tmp.json needed? Isn't it enough to save it as a variable?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+
+#*** issue_nft collection contract ***
+# issue nft
```
```suggestion
# issue a nft
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `issue_nft`
+key=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
```
```suggestion
expected_key='"_contract_address"'
expected_val='"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
same as L27-28

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** mint_nft collection contract ***
+# mint_nft
+fnsad tx wasm execute link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8 '{"mint_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","to":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","params":[{"token_type":"10000001","name":"nft1_name1","meta":"nft1_meta1"},{"token_type":"10000001","name":"nft1_name2","meta":"nft1_meta2"}]}}' --from li ...
```
same as L19

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+
+#*** mint_nft collection contract ***
+# mint_nft
```
```suggestion
# mint a nft
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `mint_nft`
+key=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
```
```suggestion
expected_key='"_contract_address"'
expected_val='"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
same as L27-28

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+
+#*** send_nft collection contract ***
+# send_nft
```
where to send?

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** send_nft collection contract ***
+# send_nft
+fnsad tx wasm execute link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8 '{"send_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","to":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","token_ids":["1000000100000001"]}}' --from link146asaycmtydq45kxc8evntqfgepagygelel00h --keyring-backend test --node http://localhost:26658 --home=${chain_ ...
```
same as L19

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `send_nft`
+key=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
```
```suggestion
expected_key='"_contract_address"'
expected_val='"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
same as L27-28

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+
+#*** burn_nft collection contract ***
+# burn_nft
```
It is the part of the above line. is this needed?

kokeshiM0chi replied:
Not needed. I deleted it.

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+
+#*** burn_nft collection contract ***
+# burn_nft
+fnsad tx wasm execute link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8 '{"burn_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","token_ids":["1000000100000001"]}}' --from link146asaycmtydq45kxc8evntqfgepagygelel00h --keyring-backend test --node http://localhost:26658 --home=${chain_dir} --chain-id simd-testing --gas 10000000  -b block -o json -y > tmp. ...
```
same as L19

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+# confirm a result of `burn_nft`
+key=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.key')
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
```
```suggestion
expected_key='"_contract_address"'
expected_val='"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"'
```

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,88 @@
+value=$(cat ./tmp.json | jq '.logs[] | .events[0]' | jq '.attributes[0]' | jq '.value')
+expected_key="\"_contract_address\""
+expected_val="\"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8\""
+test "${expected_key}" = "${key}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
+test "${expected_val}" = "${value}" ; if [ $? -eq 0 ]; then echo "ok"; else echo "ng"; fi
```
same as L27-28

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi commented:
>This PR appears to only test 4 WebAssembly files. Do you plan to test all other WebAssembly files in this PR?

This time only collection.wasm will be tested. Delete *.wasm that is not used for testing.

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
fix: use a variable `result` instead of tmp.json to check result

kokeshiM0chi committed:
fix: fix `No newline at end of file`

kokeshiM0chi committed:
chore: delete unnecessary lines

kokeshiM0chi committed:
chore: delete unnecessary lines and change `chmod 777` into `chmod 755`

kokeshiM0chi committed:
chore: delete unnecessary lines

kokeshiM0chi committed:
fix: change the function to check the result

kokeshiM0chi committed:
fix: changed to only check if message sending was successful

kokeshiM0chi committed:
fix: change to send `fnsad tx wasm execute ~` on docker instead of on local

kokeshiM0chi committed:
fix: fix contract test task with docker

kokeshiM0chi committed:
fix: remove unused wasm

kokeshiM0chi committed:
fix: remove sleep time

kokeshiM0chi committed:
fix: remove Makefile

kokeshiM0chi committed:
fix: correct to wait for processing until height becomes 1 or more

kokeshiM0chi committed:
fix: delete artifacts/checksums_intermediate.txt

kokeshiM0chi committed:
chore: delete unnecessary comments

kokeshiM0chi committed:
fix: define the valuables with values used in this test

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+set -o errexit -o nounset -o pipefail
+command -v shellcheck >/dev/null && shellcheck "$0"
```
is this needed?

kokeshiM0chi replied:
delete it

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -0,0 +1,50 @@
+source "$SCRIPT_DIR"/env
+TMP_DIR=$(mktemp -d "${TMPDIR:-/tmp}/fnsa.XXXXXXXXX")
+chmod 755 "$TMP_DIR"
+echo "Using temporary dir $TMP_DIR"
```
how about ">&2" if it is needed?

kokeshiM0chi replied:
Fixed here.
https://github.com/Finschia/finschia-wasm/pull/3/commits/f5c03a42042926bd99ff287f22c0c344a3c50256

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -0,0 +1,50 @@
+  /template/run_finschia.sh \
+  >"$FINSCHIA_LOGFILE" 2>&1 &
+
+echo "fnsad running on http://localhost:$TENDERMINT_PORT_HOST and logging into $FINSCHIA_LOGFILE"
```
how about ">&2" if it is needed?

loloicci commented on scripts/finschia/template/run_finschia.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+set -o errexit -o nounset -o pipefail
+command -v shellcheck >/dev/null && shellcheck "$0"
```
is this needed?

kokeshiM0chi replied:
delete it

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,62 @@
+URL="http://localhost:26658"
+CODE_ID=1
+
+# This is a function that checks if the result is as expected.
```
this looks simply checking the result has the word "failed"

kokeshiM0chi replied:
Fixed here
[chore: modify the comment about check_result()](https://github.com/Finschia/finschia-wasm/pull/3/commits/26b66d739d7c9937a1cbb07e7499b5194963ba15)

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,62 @@
+# This is a function that checks if the result is as expected.
+check_result() {
+    if [[ "$result" == *"failed"* ]]; then
+      exit 1
```
how about echoing why it fails to STDERR before exit?

kokeshiM0chi replied:
fixed here.

[fix: change check_run_info to receive a raw_log argument](https://github.com/Finschia/finschia-wasm/pull/3/commits/6885273a1dba397ca90f3288a62c61944be84c2b)

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,62 @@
+
+# This is a function that checks if the result is as expected.
+check_result() {
+    if [[ "$result" == *"failed"* ]]; then
```
I doubt is it enough. How about checking as before https://github.com/Finschia/finschia-wasm/pull/3/commits/8375ecb610bccdf87033878732507940ae368afe?

FYI
If define check_result as

```
check_result() {
    if [[ "$1" != "A" || "$2" != "B" ]]; then echo "err with $3"; exit 1; fi
}
```

,

```
check_result A B unexpected
```

and

```
check_result A C expected
```

work well.

kokeshiM0chi replied:
Fixed as pointed out.


[fix: change check_run_info to receive a raw_log argument](https://github.com/Finschia/finschia-wasm/pull/3/commits/6885273a1dba397ca90f3288a62c61944be84c2b)

kokeshiM0chi replied:
[fix: change check_run_info to receive a raw_log argument](https://github.com/Finschia/finschia-wasm/pull/3/commits/6885273a1dba397ca90f3288a62c61944be84c2b)

kokeshiM0chi committed:
fix: correct the loop requesting height1 to timeout in 60 seconds.

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,65 @@
+}
+
+#*** store collection contract ***
+result=$(fnsad tx wasm store ${artifacts_path}/collection.wasm --node ${URL} --from ${FROM_ACCOUNT} --chain-id simd-testing --keyring-backend test --gas 10000000 -b block -o json -y)
```
`--chain-id simd-testing --keyring-backend test -b block -o json -y` 
You can set these parameters as one const, and use it easily.

kokeshiM0chi replied:
Thanks. Fixed as pointed out.
https://github.com/Finschia/finschia-wasm/pull/3/commits/0780d9f07c9811cafcf7332984792c038d1b63ee

kokeshiM0chi committed:
fix: modify check_result

kokeshiM0chi committed:
fix: add the line at the end of this file

kokeshiM0chi commented:
@170210 san
~As pointed out by @loloicci san, if the message is sent successfully, check only the value of the event attribute.
https://github.com/Finschia/finschia-wasm/pull/3#discussion_r1467163989~

As you pointed, I checked the contents of raw_log to see if the submessage was successful. Thanks.

[fix: change check_run_info to receive a raw_log argument](https://github.com/Finschia/finschia-wasm/pull/3/commits/6885273a1dba397ca90f3288a62c61944be84c2b)

kokeshiM0chi committed:
fix: add chain_option when send transaction

kokeshiM0chi committed:
fix: change check_result to output error message

kokeshiM0chi committed:
chore: delete unnecessary processing

kokeshiM0chi committed:
chore: change to output finschia log only when its error occurs and delete unnecessary processing

kokeshiM0chi committed:
chore: modify the comment about check_result()

loloicci commented on scripts/finschia/start.sh:
```diff
@@ -38,7 +37,7 @@ docker run --rm \
   "$REPOSITORY:$VERSION" \
   /template/run_finschia.sh \
-  >"$FINSCHIA_LOGFILE" 2>&1 &
+  >"$FINSCHIA_LOGFILE" >&2 &
```
with this, no ~log~ log for stderr are issued to the log file. Isn't it?

kokeshiM0chi replied:
Yes, it is. Log issued only when an error occur

loloicci replied:
I mean I think this change is not needed and `>"$FINSCHIA_LOGFILE" 2>&1 &` works well. Sorry for confusing comment.

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -31,7 +31,7 @@ key=$(echo "${result}" | jq '.logs[] | .events' | jq '.[0]' | jq '.attributes[2]
 expected_key='"module"'
 expected_val='"wasm"'
-check_result "${key}" "${value}" "expected"
+check_result "${key}" "${value}" "${result}"
```
How about adding info on which step fails, for example `"store: ${result}"`?

kokeshiM0chi replied:
fixed here
[chore: add info on which step fails to error message of check_result](https://github.com/Finschia/finschia-wasm/pull/3/commits/0549f9e2f84abb935aaf2c49d92f3884797d68e3)

kokeshiM0chi committed:
chore: add info on which step fails to error message of `check_result`

kokeshiM0chi committed:
fix: change check_result into check_run_info which confirm if submessage is successful

kokeshiM0chi committed:
fix: redirect standard output and standard error output to variables without distinguishing between them

kokeshiM0chi committed:
fix: change `check_run_info` to receive a `raw_log` argument

kokeshiM0chi committed:
fix: change to standard output only when error occurs

kokeshiM0chi committed:
fix: change to standard output only when error occurs

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+# If the timeout fails, the process will be terminated abnormally.
+exitstatus=$?
+if [[ $exitstatus -ne 0 ]]; then
+    echo "timeout failed"
```
```suggestion
    echo "waiting finschia node is timeout"
```

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
fix: exclude artifacts

kokeshiM0chi committed:
chore: exclude artifacts

kokeshiM0chi committed:
fix: add `>&2` when echo

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+check_run_info "${raw_log}" "store: "
+
+#*** instantiate collection contract ***
+result=$(fnsad tx wasm instantiate 1 '{"name":"collection_name","uri":"collection_uri","meta":"collection_meta", "owner":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"}' --label collection1 --admin link146asaycmtydq45kxc8evntqfgepagygelel00h  --from link146asaycmtydq45kxc8evntqfgepagygelel00h --node ${URL} ${CHAIN_OPTION})
```
Please use jq to change `link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8` to `CONTRACT_ADDRESS`

kokeshiM0chi replied:
The contents of json are difficult to expand into variables. It will be difficult to return it, so please leave it as is.

kokeshiM0chi replied:
The contents of json are difficult to expand into variables. It will be difficult to return it, so please leave it as is.

loloicci replied:
FYI:

```
plus=+
echo 'foo'$plus'bar'
```

issues `foo+bar`

170210 replied:
```
INIT_MSG=$(jq -nc --arg addr "$CONTRACT_ADDRESS" \
            '{"name":"collection_name","uri":"collection_uri","meta":"collection_meta", "owner”:$addr}’ 
result=$(fnsad tx wasm instantiate 1 $INIT_MSG  --label collection1 --admin ...  --from ... --node ${URL} ${CHAIN_OPTION})
```

kokeshiM0chi replied:
fixed here
[fix: change address to variable specification instead of hard code in…](https://github.com/Finschia/finschia-wasm/pull/3/commits/409bc9e4f57487615926518589c1c78926607913)
[fix: fix the omission of address hard code](https://github.com/Finschia/finschia-wasm/pull/3/commits/a05bf66ecf4b258849f8d423b2e4c0bbd04cfa52)

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+check_run_info "${raw_log}" "instantiate: "
+
+#*** issue_nft collection contract ***
+result=$(fnsad tx wasm execute ${CONTRACT_ADDRESS} '{"issue_nft":{"name":"nft1_name","meta":"nft1_meta","owner":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8"}}' --from ${FROM_ACCOUNT} --node ${URL} --gas 10000000 ${CHAIN_OPTION})
```
ditto

kokeshiM0chi replied:
The contents of json are difficult to expand into variables. It will be difficult to return it, so please leave it as is.

kokeshiM0chi replied:
ditto

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+check_run_info "${raw_log}" "execute; issue_nft: "
+
+#*** mint_nft collection contract ***
+result=$(fnsad tx wasm execute ${CONTRACT_ADDRESS} '{"mint_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","to":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","params":[{"token_type":"10000001","name":"nft1_name1","meta":"nft1_meta1"},{"token_type":"10000001","name":"nft1_name2","meta":"nft1_meta2"}]}}' --from ${FROM_ACCOUNT} --node ${URL} --gas 1 ...
```
ditto

kokeshiM0chi replied:
The contents of json are difficult to expand into variables. It will be difficult to return it, so please leave it as is.

kokeshiM0chi replied:
ditto

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+check_run_info "${raw_log}" "execute; mint_nft: "
+
+#*** send_nft collection contract ***
+result=$(fnsad tx wasm execute ${CONTRACT_ADDRESS} '{"send_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","to":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","token_ids":["1000000100000001"]}}' --from ${FROM_ACCOUNT} --node ${URL} --gas 10000000 ${CHAIN_OPTION})
```
ditto

kokeshiM0chi replied:
The contents of json are difficult to expand into variables. It will be difficult to return it, so please leave it as is.

kokeshiM0chi replied:
ditto

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,72 @@
+check_run_info "${raw_log}" "execute; send_nft"
+
+#*** burn_nft collection contract ***
+result=$(fnsad tx wasm execute ${CONTRACT_ADDRESS} '{"burn_nft":{"from":"link14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sgf2vn8","token_ids":["1000000100000001"]}}' --from ${FROM_ACCOUNT} --node ${URL} --gas 10000000 ${CHAIN_OPTION})
```
ditto

kokeshiM0chi replied:
ditto

kokeshiM0chi committed:
fix: change address to variable specification instead of hard code in json

kokeshiM0chi committed:
fix: fix the omission of address hard code

kokeshiM0chi committed:
feat: add collection contract

kokeshiM0chi committed:
fix: add necessary files for examples/contracts

missing upload

kokeshiM0chi committed:
fix: Get code_id instead of fixed value

kokeshiM0chi committed:
fix: use only tempolary directory

kokeshiM0chi committed:
chore: remove unnecessary files (osmosis)

da1suk8 commented on scripts/finschia/run_finschia.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+echo "hello run_finschia.sh"
```
Do I need this?

da1suk8 commented:
Please note the reason for using the collection contract in the description.

kokeshiM0chi committed:
fix: change CONFIG_DIR to use only temporary folder

Prevent startup if temporary folder is not set

kokeshiM0chi committed:
fix: delete osmosis files

kokeshiM0chi committed:
fix: fix waiting state processing

kokeshiM0chi committed:
style: rename file name

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -7,7 +7,8 @@ URL='http://localhost:26658'
 
 ## Wait until the height of the block is greater than or equal to 1.
-timeout 60 bash -c "until fnsad query block 1 --node ${URL} --chain-id simd-testing > /dev/null; do sleep 0.5; done"
+timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r .result.sync_info.latest_block_height) -ge 1 ]]; do sleep 2; done'
```
```suggestion
timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r '.result.sync_info.latest_block_height // 0') -ge 1 ]]; do sleep 0.5; done'
```

loloicci reviewed (commented):
Rename "scripts/set_config_and_start_docker.sh" and "scripts/integration_test_contracts.sh" to show each does an integration test for the collection contract.

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

loloicci commented on scripts/integration_test_contracts.sh:
```diff
@@ -7,7 +7,7 @@ URL='http://localhost:26658'
 
 ## Wait until the height of the block is greater than or equal to 1.
-timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r .result.sync_info.latest_block_height) -ge 1 ]]; do sleep 2; done'
+timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r '.result.sync_info.latest_block_height // 0') -ge 1 ]]; do sleep 0.5; done'
```
```suggestion
timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r ".result.sync_info.latest_block_height // 0") -ge 1 ]]; do sleep 0.5; done'
```

loloicci reviewed (commented):
sorry for causing error

kokeshiM0chi committed:
Update scripts/integration_test_contracts.sh

Co-authored-by: TAKASE Ryo <loloicci@loloicci.dev>

kokeshiM0chi committed:
fix: remove the unused file

kokeshiM0chi commented:
>Rename "scripts/set_config_and_start_docker.sh" and "scripts/integration_test_contracts.sh" to show each does an integration test for the collection contract.

I deleted scripts/set_config_and_start_docker.sh. This is unnecessary.
Added the file "start_integration_test_contracts.sh" which execute "scripts/integration_test_contracts.sh".
@loloicci

170210 commented on examples/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+documentation = "https://docs.cosmwasm.com"
+
+[workspace.dependencies]
+cosmwasm-schema  = "1.1.9"
+cosmwasm-std     = "1.1.9"
+cosmwasm-storage = "1.1.9"
+cw-storage-plus  = "1.1.0"
+cw2              = "1.1.2"
+finschia-std     = { path = "../packages/finschia-std" }
+schemars         = "0.8.12"
+serde            = { version = "1.0.167", default-features = false, features = ["derive"] }
+thiserror        = "1.0.43"
```
Is it necessary to use Finschia package to instead.
https://github.com/Finschia/cw-workspace/blob/main/Cargo.toml

kokeshiM0chi replied:
This will be addressed in another PR.
https://github.com/Finschia/finschia-wasm/issues/6

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,74 @@
+check_run_info() {
+  local res="$1"
+  local msg="$2"
+	if [[ "$res" == *"failed"* ]]; then
+		echo -e "$msg$res" >&2
+		exit 1
+	fi
```
format

170210 commented on scripts/integration_test_contracts.sh:
```diff
@@ -0,0 +1,74 @@
+URL='http://localhost:26658'
+CHAIN_OPTION='--chain-id simd-testing --keyring-backend test -b block -o json -y'
+
+## Wait until the height of the block is greater than or equal to 1.
+timeout 60 bash -c 'until [[ $(curl -s "http://localhost:26658/status?" | jq -r ".result.sync_info.latest_block_height // 0") -ge 1 ]]; do sleep 0.5; done'
```
Is it duplicated?
```
timeout 60 bash -c "until docker ps | awk '{print $NF}' | grep finschia-wasm-testing-app > /dev/null; do sleep 0.5; done"
```

kokeshiM0chi replied:
No. Both the docker wait state and the finschia app wait state are required.
